### PR TITLE
Add installation and packaging targets

### PR DESCRIPTION
**Pull request by Congpy**
```
docker run --rm -it --mount type=bind,source=C:\Users\goncb\Classes\cpack-exercise-wt2223,target=/mnt/cpack-exercise 3ea6df4c227abe87121b7e7866f84862e744450d35b0428cc9e67b76ed56f769
apt install ./build/cpackexample_0.1.0_amd64.deb
cpackexample
```

image

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -21,8 +21,42 @@ find_package(yaml-cpp
...
+install(TARGETS "${PROJECT_NAME}" cpackexamplelib
+	LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+	ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
+	RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
+	PUBLIC_HEADER DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/cpackexamplelib"
+  	INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/cpackexamplelib"
+  )
+
+include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/CPackConfig.cmake)
+include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/Packaging.cmake)
```
Comment (uekerman) [inline CMakeLists.txt:60]: Why two different files? Including `CPack` multiple times might also give problems.

File: cmake/Packaging.cmake [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Debian packaging
+set(CPACK_GENERATOR "TGZ;DEB")
```
Comment (uekerman) [inline cmake/Packaging.cmake:2]: This did not work. Probably because you included `CPack` multiple times.

**Review by uekerman — COMMENTED** [review]
Please don't add `build` folders :grin: 

I was not able to create a Debian package:

```
root@e857fe571ada:/mnt/cpack-exercise/build# make package && dpkg --info cpackexample_0.1.0_amd64.deb && ls
Consolidate compiler generated dependencies of target cpackexamplelib
[ 71%] Built target cpackexamplelib
Consolidate compiler generated dependencies of target cpackexample
[100%] Built target cpackexample
Run CPack packaging tool...
CPack: Create package using TBZ2
CPack: Install projects
CPack: - Install directory: /mnt/cpack-exercise
CPack: Create package
CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.bz2 generated.
CPack: Create package using TGZ
CPack: Install projects
CPack: - Install directory: /mnt/cpack-exercise
CPack: Create package
CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.gz generated.
CPack: Create package using TXZ
CPack: Install projects
CPack: - Install directory: /mnt/cpack-exercise
CPack: Create package
CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.xz generated.
CPack: Create package using TZ
CPack: Install projects
CPack: - Install directory: /mnt/cpack-exercise
CPack: Create package
CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.Z generated.
dpkg-deb: error: failed to read archive 'cpackexample_0.1.0_amd64.deb': No such file or directory
root@e857fe571ada:/mnt/cpack-exercise/build# 
root@e857fe571ada:/mnt/cpack-exercise/build# ls
CMakeCache.txt           Makefile             cpackexample-0.1.0-Source.tar.Z    install_manifest.txt
CMakeFiles               _CPack_Packages      cpackexample-0.1.0-Source.tar.bz2  libcpackexamplelib.so
CPackConfig.cmake        cmake_install.cmake  cpackexample-0.1.0-Source.tar.gz   solution.vtk
CPackSourceConfig.cmake  cpackexample         cpackexample-0.1.0-Source.tar.xz
```

I am closing this PR to maintain a good overview of all submissions.

File: CMakeLists.txt (hunk 1 above)
Comment (Congpy) [reply CMakeLists.txt:60]: I thought maybe CPack keeps track of inclusions and handles them properly even though they are mentioned multiple times. Should I correct it?

File: cmake/Packaging.cmake (hunk 2 above)
Comment (Congpy) [reply cmake/Packaging.cmake:2]: Interesting, I did run a test run this time.

**Comment by Congpy** [comment]
> Please don't add `build` folders 😁
> 
> I was not able to create a Debian package:
> 
> ```
> root@e857fe571ada:/mnt/cpack-exercise/build# make package && dpkg --info cpackexample_0.1.0_amd64.deb && ls
> Consolidate compiler generated dependencies of target cpackexamplelib
> [ 71%] Built target cpackexamplelib
> Consolidate compiler generated dependencies of target cpackexample
> [100%] Built target cpackexample
> Run CPack packaging tool...
> CPack: Create package using TBZ2
> CPack: Install projects
> CPack: - Install directory: /mnt/cpack-exercise
> CPack: Create package
> CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.bz2 generated.
> CPack: Create package using TGZ
> CPack: Install projects
> CPack: - Install directory: /mnt/cpack-exercise
> CPack: Create package
> CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.gz generated.
> CPack: Create package using TXZ
> CPack: Install projects
> CPack: - Install directory: /mnt/cpack-exercise
> CPack: Create package
> CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.xz generated.
> CPack: Create package using TZ
> CPack: Install projects
> CPack: - Install directory: /mnt/cpack-exercise
> CPack: Create package
> CPack: - package: /mnt/cpack-exercise/build/cpackexample-0.1.0-Source.tar.Z generated.
> dpkg-deb: error: failed to read archive 'cpackexample_0.1.0_amd64.deb': No such file or directory
> root@e857fe571ada:/mnt/cpack-exercise/build# 
> root@e857fe571ada:/mnt/cpack-exercise/build# ls
> CMakeCache.txt           Makefile             cpackexample-0.1.0-Source.tar.Z    install_manifest.txt
> CMakeFiles               _CPack_Packages      cpackexample-0.1.0-Source.tar.bz2  libcpackexamplelib.so
> CPackConfig.cmake        cmake_install.cmake  cpackexample-0.1.0-Source.tar.gz   solution.vtk
> CPackSourceConfig.cmake  cpackexample         cpackexample-0.1.0-Source.tar.xz
> ```
> 
> I am closing this PR to maintain a good overview of all submissions.

Does this mean I need to correct them or I should just refrain mentioning them multiple times AND avoid pushing build folder for the next time? 😄

**Comment by uekerman** [comment]
No must to correct things here, was mainly meant as feedback. You can, however, modify things if you want to test again.

**Pull request closed without merging**